### ISSUE TITLE
[REVIEW]: BCImat: a Matlab-based framework for Intracortical Brain-Computer Interfaces and their simulation with an artificial spiking neural network

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3956]
**Submitting author:** @eferrea (<a href="http://orcid.org/0000-0003-1352-4789">Enrico Ferrea</a>)
**Repository:** https://github.com/eferrea/BCI-Master
**Branch with paper.md** (empty if default branch): 
**Version:** v.1.0.0
**Editor:** @oliviaguest
**Reviewers:** @mstimberg, @puolival
**Archive:** 10.5281/zenodo.6759182

 

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/b6f403bdc8f1b9c96698feca6a466169/status.svg)](https://joss.theoj.org/papers/b6f403bdc8f1b9c96698feca6a466169)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/b6f403bdc8f1b9c96698feca6a466169"><img src="https://joss.theoj.org/papers/b6f403bdc8f1b9c96698feca6a466169/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/b6f403bdc8f1b9c96698feca6a466169/status.svg)](https://joss.theoj.org/papers/b6f403bdc8f1b9c96698feca6a466169)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/eferrea/BCI-Master" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@mstimberg & @puolival, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @oliviaguest know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @mstimberg
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/eferrea/BCI-Master">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/eferrea/BCI-Master">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@eferrea) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @puolival
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/eferrea/BCI-Master">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/eferrea/BCI-Master">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@eferrea) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @mstimberg, @puolival it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 1175

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.09 s (332.1 files/s, 36526.5 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
MATLAB                          16            371            342           1071
C++                              4            216            121            488
XML                              3              0              0            176
TeX                              1             13              0            173
Markdown                         2             53              0             95
C/C++ Header                     1              9             10             20
YAML                             1              1              4             18
DOS Batch                        1              0              0              9
-------------------------------------------------------------------------------
SUM:                            29            663            477           2050
-------------------------------------------------------------------------------


Statistical information for the repository '72453a089fb461b9e3bf5986' was
gathered on 2021/11/29.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
nerofumo82                      10           918             54          100.00

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
nerofumo82                  864           94.1          0.1               15.39
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1126/science.aaa5417 is OK
- 10.1016/S0140-6736(17)30601-3 is OK
- 10.1016/S0140-6736(12)61816-9 is OK
- 10.1038/nn.3265 is OK
- 10.1038/nature11076 is OK
- 10.1073/pnas.0808113105 is OK
- 10.1038/nature10845 is OK
- 10.1038/nature13665 is OK
- 10.1016/j.neuron.2014.08.038 is OK

MISSING DOIs

- 10.1101/2021.02.16.431440 may be a valid DOI for title: Statistical determinants of visuomotor adaptation in a virtual reality three-dimensional environment
- 10.1523/jneurosci.02-11-01527.1982 may be a valid DOI for title: On the relations between the direction of two-dimensional arm movements and cell discharge in primate motor cortex
- 10.1109/thms.2020.2968411 may be a valid DOI for title: Brain–computer interface software: A review and discussion

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03956/joss.03956/10.21105.joss.03956.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03956/joss.03956/10.21105.joss.03956.pdf) :page_facing_up: :point_left:

**Comment by eferrea** [comment]
Hi @oliviaguest, I am here in the new page. Thanks @puolival for reviewing the paper.

**Comment by oliviaguest** [comment]
Hi all! 👋 Thank you so much @mstimberg, @puolival for accepting to review this. Please read the instructions above. Any questions, feedback on the paper, etc., please post here. Any very code-specific questions, suggestions, etc., please use the issues in the code repo and link to them from this thread, so we can all keep track of them. 🌸

For examples of how this process plays out feel free to skim previous reviews, such as: #2285 and #2348. ☺️

**Comment by mstimberg** [comment]
## Review @mstimberg
As mentioned earlier, I am neither a MATLAB-expert, nor do I know much about BCI – so please read my review with this in mind.
 
 I was able to install and compile all dependencies, as well the `TrackM` software and the mex files for the Matlab GUI presented in this paper (under Ubuntu Linux 20.04). I was able to run both pieces of software and test its basic functionality, and did not encounter any major problem. I do not have access to Blackrock neural interface, so I obviously could not test the package in a real BCI application.
 I have opened a number of issues on the main repository (eferrea/BCI-Master#2, eferrea/BCI-Master#3, eferrea/BCI-Master#4, eferrea/BCI-Master#5, eferrea/BCI-Master#6), with suggestions for improvements to documentation, usability and testing. I fear that in its current state, the package will be difficult to use for new users, and some parts of it look like code that is rather aimed for internal than for general use (e.g. hardcoded private IP addresses). Everything looks to be fixable, though, I did not encounter any major blocking issues.
 Regarding the Software paper, I do not have much to criticize. As friendly whedon noted earlier, three papers are missing DOIs and one of them is also missing the year (leading to a "n.d." in the citation).

**Comment by eferrea** [comment]
Hi @mstimberg, many thanks for the constructive review. I will address all of  your concerns as soon as possible.

**Comment by whedon** [comment]
:wave: @puolival, please update us on how your review is going (this is an automated reminder).

**Comment by mstimberg** [comment]
@eferrea thanks for the substantial changes, all of my concerns have been addressed (except for the few minor issues I commented in eferrea/BCI-Master#2, eferrea/BCI-Master#3, eferrea/BCI-Master#4). Before the formal acceptance from my side, there's only one more thing that I unfortunately forgot to mention earlier: it would be good to make an official release on github (or just a tag) and give the software a version number. This will make it much clearer which version of the code was described by the JOSS paper, and will allow you to easily communicate changes with respect to this version in the future. But it might make sense to wait with this until the end of the reviewing process, of course.

**Comment by mstimberg** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03956/joss.03956/10.21105.joss.03956.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03956/joss.03956/10.21105.joss.03956.pdf) :page_facing_up: :point_left:

**Comment by eferrea** [comment]
@mstimberg ,  thanks for your comments. I answered your questions (eferrea/BCI-Master#2, eferrea/BCI-Master#3, eferrea/BCI-Master#4).
In addition, I have to mention that I just added in the paper.md a small paragraph of our funding sources. 
As you suggested, I will then wait until the end of the review process to make an official release.

**Comment by mstimberg** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03956/joss.03956/10.21105.joss.03956.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03956/joss.03956/10.21105.joss.03956.pdf) :page_facing_up: :point_left:

**Comment by mstimberg** [comment]
Thanks for the additional changes @eferrea. I just noticed that the paper still has a broken reference (`[RN1]` in line 76), but apart from that I am happy with the current state of the software/paper.

**Comment by eferrea** [comment]
Many thanks @mstimberg for your review. I also corrected the broken reference now.

**Comment by mstimberg** [comment]
@whedon generate pdf

**Comment by mstimberg** [comment]
Great, so just to make it official: I recommend acceptance (CC @oliviaguest )

**Comment by oliviaguest** [comment]
@mstimberg thank you so much!

@puolival is everything clear with what you need to do to review this? 😊

**Comment by puolival** [comment]
@oliviaguest I am almost finished writing the review. I'll report back in a couple of days.

**Comment by puolival** [comment]
### Review 21.1.2022

**Manuscript**

1. The manuscript is currently too short. Please report results at least from one actual experiment. How good cursor control performance can be achieved using this software? Please also compare the performance to a typical state-of-the-art BCI experiment. Please also state more clearly the main functions of this software and how does it compare to the other existing software. Please also mention the most important limitations and/or how the software can be developed further.

2. It is unclear what is meant by the sentence “The code was successfully used from extracted neural activity of rhesus monkeys performing a similar center-out reach as previously described in a virtual-reality environment …”. If this means that experiments on rhesus monkeys were performed for the purpose of developing this software, the relevant ethical permissions, procedures, methods, etc. must be carefully reported according to standard good practices. It is not possible to recommend publication if an animal model was indeed used and these information are missing.

3. Related to points 1. and 2., is this software mainly intended to be used with recordings done on humans? This is currently unclear due to the reference to animal models. The abstract only refers to human patients. If the software is intended to be used with human patients, it must be tested with experiments involving human patients.

4. Does the statement “… movement stage information and position of the target are handled to clear the spike buffer regularly to avoid excessive memory overload” mean that the size of the buffer is dynamic during run-time? It would be much more preferable to use a constant-sized circular buffer (also known as circular queue or a ring buffer) instead, which would not have such issues.
 
5. Has this software been tested with any other acquisition system than Cereplex? The Cereplex system is now mentioned as an example but if the software is not (known to be) compatible with other acquisition systems, it should be clearly stated. The manuscript now gives an impression of a general-purpose system but if the software is in fact written for a specific hardware vendor, excluding the SNN part, this should be clearly communicated, perhaps already in the title.

6. “We also implemented some important BCI features that are frequently used in the literature to provide efficient training and better decoding performance”. How much did the performance improve? How was the performance estimated? Please report the results that were obtained.

7. Please proofread the manuscript once more. In particular, please check that words that should be in singular form are not accidentally in plural form or vice versa. Spaces are sometimes missing before citation parentheses. Check that words are capitalized correctly (e.g. MATLAB is sometimes referred to as ‘Matlab’ and sometimes as ‘MATLAB’).

8. In the reference list, journal names are sometimes written in full and sometimes abbreviated. Please adhere to a consistent style.

9. Font sizes in Figure 1 are too small. Please ensure that all text is readable when the manuscript is printed.

10. Please write all abbreviations once in full.

11. To make the manuscript more accessible, it would be good if it was explained what is a “standard c++ client-server application”.

12. The main features of the SNNs must be described (which neuron model is used, size and connectivity of the network, is the network static or is there plasticity, etc.)

**Software**

1. The license file contains the name of only one author while several authors have written the software (based on comments in the source files).

2. There is no automated package management solution, which is a JOSS recommendation. Also, the installation instructions do not contain version numbers of external packages (e.g. VRPN), which is likely to cause issues sooner or later. The user also needs to manually copy files in order to get the software running, which is a step that I think should be automated.

3. There are no automated tests. The files “TestBlackrock.m” and “TestVRPNConnection.m” (neither of which is properly commented) can be used to test hardware/internet connections but there are no tests for assessing functionality of this software. The authors should write at least unit tests for all functions.

4. Many functions and methods are very long, might do several things, and are often not easy to understand. I suggest that long functions are refactored into several smaller ones, such that each function does one only thing (i.e. adherence to the so-called keep-it-simple principle).

5. There are comments such as “… many of the parameters listed below are not used with the basic provided task controller but were used during our experiments”. It therefore appears that this software might not yet be suitable for general use, since it has not been written in a general-purpose context.

6. All functions, their input/output arguments, and non-trivial block of code should be properly commented. Indentation should be uniform across files. Single-letter variable names should not be used unless they are part of a mathematical equation.

7. Adhering to a specific style guide (see e.g. https://google.github.io/styleguide/) would make it easier to read the program code.

8. The software would be more valuable for researchers if it was easy to customize the parameters of the spiking neural network, for example via a separate GUI.

9. Like the first reviewer, I am not able try this out in a real application due to not having access to the kind of hardware that was used by the authors. Ideally, a third reviewer would assess performance in a similar context as the original one.

**Summary**

The manuscript and software must be both substantially expanded, and another review is required after the necessary changes have been made. Currently, the work appears to be more of a minor side product of other scientific work rather than a complete standalone software. This said, the software does have good potential to become a useful tool for the BCI community. While I am currently not able to recommend publication, I do strongly encourage the authors to continue development.

**Comment by puolival** [comment]
@oliviaguest It seems that I am not able to edit the checklist even though I am logged in.

**Comment by eferrea** [comment]
Hi @puolival (and hi @oliviaguest), many thanks for your review. 
It seems I can fix all of the reported issues.
 
Please bear in mind that the software was a necessary requirement (and NOT a minor side product) to perform BCI studies in our non-human primate labs and other different studies are now arising from this software.  In addition, to the best of our knowledge, there is no software grouping together all the most important and necessary features for conducting intracortical BCI experiments in non-human primates, and despite some labs might have developed their own ad hoc solutions, scientists that want to start in this field might consider the software as a valuable option. Indeed please consider, that many important details of the methodological implementation are often not reported in scientific papers but need to be inferred from direct experience. Therefore, this software was developed and optimized under several years of direct experimental testing.

**Comment by oliviaguest** [comment]
@puolival I think you need to be reinvited by @whedon because you didn't accept the last time. No problem tho!

**Comment by oliviaguest** [comment]
@whedon add @puolival as reviewer

**Comment by danielskatz** [comment]
@whedon remove @puolival as reviewer

We just need each reviewer once :)

**Comment by whedon** [comment]
OK, @puolival is no longer a reviewer

**Comment by danielskatz** [comment]
@whedon re-invite @puolival as reviewer

this is the way to get a new invitation...

**Comment by whedon** [comment]
The reviewer already has a pending invite.

@puolival please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by oliviaguest** [comment]
Ooops, thanks @danielskatz!

**Comment by danielskatz** [comment]
If that link fails, tell me and I'll reissue it - there's a timing bug that sometimes happens here.

**Comment by danielskatz** [comment]
ok, @whedon didn't do what I wanted and removed both @puolival's - let's add one back

**Comment by danielskatz** [comment]
@whedon add @puolival as reviewer

**Comment by whedon** [comment]
OK, @puolival is now a reviewer

**Comment by danielskatz** [comment]
Whedon and the issue now are correct, so we just need to make sure @puolival is able to check off items by accepting the invitation, and if they cannot, we need to reissue it one more time, then it should work

**Comment by oliviaguest** [comment]
@danielskatz thank you!

**Comment by puolival** [comment]
@eferrea Thanks for the detailed response. Where can I access the updated manuscript (pdf)?

**Comment by mstimberg** [comment]
To get the rendered PDF here we should also be able to use

@whedon generate pdf

**Comment by mstimberg** [comment]
@whedon generate pdf

[sorry, keep forgetting that whedon commands need to start a comment]

**Comment by eferrea** [comment]
Hi @puolival and (hi @oliviaguest), thanks again for reviewing the manuscript. Since we have a manuscript and a grant pending on this paper I would ask if a much longer time is needed for the review so that in case we will go ahead with the other submissions. Please let me also know if meanwhile, I could provide additional information to agevolate the process. Thanks again for your work and best regards.

**Comment by mstimberg** [comment]
Hi everyone. In case there are any doubts about this: I still recommend the acceptance of the paper in its most recent version.
Quite obviously @puolival's review was  more "demanding" than my own, but I also agree that the changes to address the raised points improved the manuscript/code/documentation.

I hope I am not overstepping my role as a reviewer here, but let me mention two points from the JOSS guidelines that might affect the evaluation of the current manuscript and code:

> Your paper must not focus on new research results accomplished with the software.

([Submission guidelines for authors](https://joss.readthedocs.io/en/latest/submitting.html)) I think it is important to keep in mind that the role of papers in JOSS is different from a "software paper" in most other journals.

> Authors are strongly encouraged to include an automated test suite covering the core functionality of their software.
>
>    Good: An automated test suite hooked up to continuous integration (GitHub Actions, Circle CI, or similar)
>    **OK: Documented manual steps that can be followed to objectively check the expected functionality of the software (e.g., a sample input file to assert behavior)**

([Review criteria: tests](https://joss.readthedocs.io/en/latest/review_criteria.html#tests)) While I of course agree that unit tests are important, I also acknowledge that the core part of the presented work involves GUI and client-server communication, which means that a fully automated test suite presents a considerable effort.

**Comment by oliviaguest** [comment]
@eferrea let me know if you would like to pause this submission or if you just would like to work on it slowly but within a sensible timeframe. Also, please feel free as I think @mstimberg is hinting to triage the feedback you have into categories including ones you do want to do or cannot do, and we can discuss such feedback that might not be possible for you to do. Does this make sense?

**Comment by eferrea** [comment]
@oliviaguest , after my last submission I did not receive any feedback from @puolival. If there are still open issues that are necessary to accept the paper I am happy to address them.

**Comment by oliviaguest** [comment]
@eferrea I want to give a chance to the 2nd reviewer to get back to us. @puolival do you have a rough ETA in mind for checking this? ☺️

**Comment by puolival** [comment]
@oliviaguest @eferrea

Thank you for the revised manuscript and software. Overall, I think this software has great potential to become a very useful tool for the BCI community. However, while some improvements have been now made after the initial submission, many of my previous major comments have not been addressed. Therefore, I cannot recommend accepting this submission at this stage.

I have here summarized part of the issues that I already pointed out in the previous review:

1. The manuscript is too underdeveloped. There is no full description of the performed experiments or the obtained results. For example, the manuscript does not contain data that supports the claim "To demonstrate functionality and performance, the BCImat code was successfully used with neural activity recorded in motor cortical areas of two rhesus monkeys performing a center-out reach task in a virtual-reality environment, .." What exactly were those experiments and how were they conducted? What were the results? Since no complete data from an actual experiment has been reported, it is impossible to tell whether the software and methods actually work or not. Information about the ethical approvals has been added, which is good.

2. Evidence for the translational value must be added to the manuscript. If this software is intended to be used in recordings done on humans, there must be data that shows the software works in such context.

3. There are no comparisons to other BCI software.

4. There is still no automated package management solution and installing this software still requires substantial effort on the end user's part: locating, downloading, and installing external libraries, compiling the task controller, and editing environmental variables and configuration files. Ideally, there would be just one installation script that the user needs to run, or an equivalent simple solution. Compare the situation for example to installing Python software libraries, where something like "pip install bcimat" could be sufficient in an ideal case.

5. Automated tests have not been added. I agree with @mstimberg that writing some of these tests could require considerable effort but this is a JOSS requirement. Also, even limited testing would be much better than no testing at all.

6. My previous comment "Many functions and methods are very long, might do several things, and are often not easy to understand. I suggest that long functions are refactored into several smaller ones, such that each function does one only thing" has not been completely addressed. Input/output variables remain uncommented in many parts. Like previously, it is still on many occassions not possible to understand what the variables are for, what are the allowed data/value ranges, etc. Some variables have been renamed and comments added but for the most part the software has not been revised for greater clarity.

7. The drawbacks of having no connectivity between the neurons in the SNN should be acknowledged. This makes the simulation biophysically very unrealistic and there probably can't be population-level phenomena typical to real recordings, e.g. oscillations in the population activity. The claim "Importantly, it also allows realistic real-time neural data simulations .." in the abstract is now too strong. No additional GUI was implemented for adjusting the parameters of the SNN.

8. Please add the required version numbers also for MATLAB and the Statistics and Machine Learning Toolbox.

Finally, I want to emphasize that all comments are intended as constructive criticism. In my opinion, addressing the issues pointed out here and in the previous review would make it much easier to adopt this software into one's use and/or help contribute to the software.

**Comment by mstimberg** [comment]
My earlier comment probably made clear that I do not quite share @puolival's assessment, in particular that the "manuscript is underdeveloped" – as @eferrea pointed out, the manuscript is already longer than recommended. But of course my opinion shouldn't have any more weight here than anyone else's.

I just wanted to comment on one point that referred to my earlier comment:

> 5\. Automated tests have not been added. I agree with @mstimberg that writing some of these tests could require considerable effort but this is a JOSS requirement. Also, even limited testing would be much better than no testing at all.

Automated tests are "strongly encouraged" by JOSS, but they are not a requirement. The [review criteria](https://joss.readthedocs.io/en/latest/review_criteria.html) explicitly state that "Documented manual steps that can be followed to objectively check the expected functionality" are considered "OK" .

**Comment by puolival** [comment]
@mstimberg Thank you for clarifying this. If my reading of the guidelines was too strict, I am happy to continue as you see best @oliviaguest.

**Comment by oliviaguest** [comment]
Thanks, all. 

I'm going to think about some points raised above, but I do think 4, 6, 8, pasted below, are worthy of comment/consideration from the authors: 

> 4. There is still no automated package management solution and installing this software still requires substantial effort on the end user's part: locating, downloading, and installing external libraries, compiling the task controller, and editing environmental variables and configuration files. Ideally, there would be just one installation script that the user needs to run, or an equivalent simple solution. Compare the situation for example to installing Python software libraries, where something like "pip install bcimat" could be sufficient in an ideal case.

> 6. My previous comment "Many functions and methods are very long, might do several things, and are often not easy to understand. I suggest that long functions are refactored into several smaller ones, such that each function does one only thing" has not been completely addressed. Input/output variables remain uncommented in many parts. Like previously, it is still on many occassions not possible to understand what the variables are for, what are the allowed data/value ranges, etc. Some variables have been renamed and comments added but for the most part the software has not been revised for greater clarity.

> 8. Please add the required version numbers also for MATLAB and the Statistics and Machine Learning Toolbox.

**Comment by eferrea** [comment]
Dear Editor (@oliviaguest), many thanks for restricting the review to the most reasonable points. We see the benefit in addressing point 6, which can be achieved with reasonable effort given that the code is also already shared inside the lab and with external collaborators. Point 8 is very minor and we have already addressed it in the readme file. Regarding point 4, we were wondering whether in its actual version installation instructions are not already satisfying JOSS guidelines as we read them at https://joss.readthedocs.io/en/latest/review_checklist.html since automated package management solution is not a strict requirement and a list of dependencies is clearly-stated:
  
* Installation instructions: Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.

We appreciate your time and effort in considering our manuscript and look forward to your feedback.
Kind regards, Enrico Ferrea (on behalf of all authors)

**Comment by oliviaguest** [comment]
@eferrea I suggest taking your time to address the parts you can, and we can discuss it once you have. I am totally happy with you *not using* a package management solution, as long as the steps are clear, indeed. 😊

**Comment by eferrea** [comment]
@puolival , @oliviaguest , @mstimberg 

Thanks for taking the time to review our software. Here we address/discuss at our best the remaining points.

4. There is still no automated package management solution and installing this software still requires substantial effort on the end user's part: locating, downloading, and installing external libraries, compiling the task controller, and editing environmental variables and configuration files. Ideally, there would be just one installation script that the user needs to run, or an equivalent simple solution. Compare the situation for example to installing Python software libraries, where something like "pip install bcimat" could be sufficient in an ideal case.

So far, we implemented @mstimberg as well as @puolival suggestions to make the installation procedure simpler. We think that after these improvements, the actual version provides a clearly stated list of dependencies that can be followed to install the software which is the specific JOSS requirement.  

6. My previous comment "Many functions and methods are very long, might do several things, and are often not easy to understand. I suggest that long functions are refactored into several smaller ones, such that each function does one only thing" has not been completely addressed. Input/output variables remain uncommented in many parts. Like previously, it is still on many occasions not possible to understand what the variables are for, what are the allowed data/value ranges, etc. Some variables have been renamed and comments added but for the most part, the software has not been revised for greater clarity.

We now add missing comments documenting all input/output variables to improve understanding of what these variables do inside the functions. We specified which data ranges are used in the calibrator function (see BCI_loop.m lines 102-103). We believe that in the current version the code can be more easily read. However, we did not find excessively long functions that could be refactored into smaller ones without impacting the core of the project that is already shared inside the lab and with external collaborators.  

8. Please add the required version numbers also for MATLAB and the Statistics and Machine Learning Toolbox.
 
Used Matlab version (2016a) and toolbox version (10.2) are now specified in the readme file.

We hope that after considering the points raised by the reviewer, the manuscript now fulfills the criteria for publication.

**Comment by oliviaguest** [comment]
@puolival would you be happy to edit the checklist in the OP to reflect how you see the current version of this project? ☺️

**Comment by puolival** [comment]
@oliviaguest

I am still not able to modify the list. If I click the previous invitation link again, I get the following error message "Sorry, we couldn't find that repository invitation. It is possible that the invitation was revoked or that you are not logged into the invited account."

Anyway, as it was decided that a large part of my review will not be addressed, I think you can continue with the submission as you see best, and that further input wouldn't change this. Thanks.

**Comment by danielskatz** [comment]
@puolival - because JOSS changed its bot, some reviews that were started but where a reviewer didn't accept an invitation ended up in an odd situation.  The best option here is for you to write `@editorialbot generate my checklist` as a new comment - this will generate a new checklist for you that you can then fill in.  You can then ignore your part of the original checklist (and @oliviaguest could edit it to remove your part if she wished)

**Comment by eferrea** [comment]
Hi @oliviaguest , how should we proceed from here?
Thanks.

**Comment by oliviaguest** [comment]
@eferrea and @puolival I did not restrict the review. Let's be very clear here, what I said above in https://github.com/openjournals/joss-reviews/issues/3956#issuecomment-1073573426 was this: 

> I'm going to think about some points raised above, but I do think 4, 6, 8, pasted below, are worthy of comment/consideration from the authors:

Saying something needs to be thought about while highlighting some parts does not mean the review is restricted to only those parts. I am going to think further about this, but please do not misunderstand me.

**Comment by oliviaguest** [comment]
@puolival may I ask you to separate your review to differentiate between 'essential' (must address) feedback and 'nice to have' from your perspective, please? And then, we can see what compromises can be made here as indeed it's not always possible for reviewers and authors to agree on what essential changes are, and it is my job to do so. I would like to underline that I am very thankful, and I am sure the authors are too, for the time you have invested here to give such detailed feedback and your care and attention is greatly appreciated.

**Comment by eferrea** [comment]
Hi @puolival @oliviaguest , just a reminder that we did not receive any feedback yet. Thanks for your contribution.

**Comment by eferrea** [comment]
Dear @oliviaguest, given the actual situation, we kindly ask you how to proceed from here. So far we addressed or discussed all the main remaining issues but unfortunately we did not received any feedback yet (see https://github.com/openjournals/joss-reviews/issues/3956#issuecomment-1078829728 and https://github.com/openjournals/joss-reviews/issues/3956#issuecomment-1070967659. 
Because of that we think that in its current version ( and also as @mstimberg pointed out https://github.com/openjournals/joss-reviews/issues/3956#issuecomment-1063096197) the software is suitable for publication.

**Comment by oliviaguest** [comment]
@eferrea please be a little patient with @puolival — if they do not get back to us within the next few weeks, we can either find another 2nd reviewer or I will decide what to do from there. I hope that is OK. ☺️

**Comment by eferrea** [comment]
Hi @oliviaguest, many thanks for your answer . From our side is fine to be patient (we also were already :) ), however we would like to know whether @puolival is still willing to finish the review given that he did not replied any messages since before Easter. If he wants to continue we can wait, otherwise we think that would be better to take an alternative decision now. Maybe he could shortly comment on that.

**Comment by eferrea** [comment]
Dear @oliviaguest, at this point we wanted to solicit a decision upon our manuscript. On March 30th reviewer 2 said “Anyway, […], I think you can continue with the submission as you see best, and that further input wouldn't change this. Thanks.”. From our side it is obvious that reviewer 2 clearly indicated with this statement that they have no intention to continue with the review; consequently, no further messages followed this statement since 2 months.  Further patience for “few weeks”, as you suggest, will almost certainly not change this situation. 

From our side, we answered or discussed all the raised points in our previous reply https://github.com/openjournals/joss-reviews/issues/3956#issuecomment-1078829728. To summarize, we added comments to the code for better clarity and we modified the text to fulfill points 6,8 of comment https://github.com/openjournals/joss-reviews/issues/3956#issuecomment-1070967659. We discussed that points 4-5 are already satisfied according to journal guidelines, whereas  point 1  was already addressed in our previous revision by adding information on average target acquisition time as an effective measure of BCI performance. As already mentioned, we could add several and similar measurements but this will add additional text to the paper without adding additional important information. We also discuss that our software for intracortical BCI cannot really be compared with available EEG based software given the very different types of physiological signals. We think that at this stage, the main points from two reviewers are largely satisfied. We therefore would kindly ask you to accept the paper in its current version to avoid further delay since we are by now 9 months after first submission. 
Enrico Ferrea on behalf of all authors

**Comment by danielskatz** [comment]
👋 @eferrea and @oliviaguest - I'm writing as the JOSS Associate-Editor-in-Chief on duty this week.

I apologize to @eferrea for the length of time this is taking.

However, peer review does not succeed by time passing.  I think @oliviaguest needs to find another reviewer who can carry our a review quickly if possible, or perhaps step in and perform the review herself if she has time.

**Comment by oliviaguest** [comment]
Agreed. Thanks @danielskatz. Do you have any suggestions, @eferrea? I'm happy to go with who you might have in mind.

**Comment by oliviaguest** [comment]
I don't have time sadly. It would have to wait till after September. I'm certain, that's too long and unnecessarily so.

**Comment by danielskatz** [comment]
@oliviaguest - Having re-read this thread, I think you can make a decision if you think that the comments from @puolival's review needed to make this acceptable to JOSS have been met, in the absence of further feedback from @puolival.

Alternatively, you can try to find a new reviewer.

If you think I've missed an option, please let me know.

**Comment by oliviaguest** [comment]
No, indeed, I have more time now (Dutch teaching goes till the end of July sadly) — and I am happy to recommend that we proceed without a new reviewer.

**Comment by oliviaguest** [comment]
@editorialbot check references

**Comment by oliviaguest** [comment]
@eferrea can you fix the DOI issue, please? 😊

**Comment by eferrea** [comment]
Hi @oliviaguest , thanks for accepting finishing reviewing. I corrected the DOI.

**Comment by oliviaguest** [comment]
@editorialbot check references

**Comment by oliviaguest** [comment]
@editorialbot check repository

**Comment by oliviaguest** [comment]
@eferrea can you please create an archive for your software (on [Zenodo](https://zenodo.org/), [figshare](https://figshare.com/), or other) and post the archive DOI here?

**Comment by eferrea** [comment]
@oliviaguest thanks for your message, I archived the software on Zenodo [DOI: 10.5281/zenodo.6759182] . I also updated the affiliations and updated bibliography with one paper that was on the bioRxiv and recently published.

**Comment by oliviaguest** [comment]
@editorialbot check references

**Comment by oliviaguest** [comment]
@editorialbot generate pdf

**Comment by oliviaguest** [comment]
@editorialbot set 10.5281/zenodo.6759182 as archive

**Comment by oliviaguest** [comment]
Also, please review and merge https://github.com/eferrea/BCI-Master/pull/7, @eferrea — just some typos.

**Comment by mstimberg** [comment]
I certainly don't intend to complicate things and to draw out the publishing process further, but I wonder: other journals I've reviewed for that mention the reviewer names in the paper state something along the lines of "reviewers endorsed the manuscript in its current form", although I'm not sure if that's the case for JOSS? I'm saying this since I'm not 100% sure that @puolival would subscribe to that statement and might not be happy to be listed on the paper.

**Comment by eferrea** [comment]
Hi @oliviaguest, thanks for your edits. I merged them and then I corrected affiliation number 4 since it has recently changed. I also agree with @mstimberg that @puolival should not be listed as reviewer of the paper.

**Comment by oliviaguest** [comment]
@eferrea sorry, but why? Didn't their contribution get used? This is a strange request unless you think their labour was worth nothing to you.

**Comment by eferrea** [comment]
@oliviaguest for me it is completely fine if the JOSS regulations are fine with that. Thanks.

**Comment by mstimberg** [comment]
@oliviaguest (sorry, forgot to tag you earlier): to avoid misunderstandings, @eferrea's comment was related to my earlier comment:
> I certainly don't intend to complicate things and to draw out the publishing process further, but I wonder: other journals I've reviewed for that mention the reviewer names in the paper state something along the lines of "reviewers endorsed the manuscript in its current form", although I'm not sure if that's the case for JOSS? I'm saying this since I'm not 100% sure that @puolival would subscribe to that statement and might not be happy to be listed on the paper.

I am completely convinced that @puolival *deserves* to be listed to be a reviewer, I am just not 100% sure that they *want* to be listed as a reviewer.

**Comment by eferrea** [comment]
@oliviaguest , sorry I also did not notice you were not tagged.

**Comment by oliviaguest** [comment]
I appreciate you raising this, but I don't appreciate being told, even if indirectly, how to manage this. I'll happily give @puolival time to reply, but I'm wondering if maybe an @openjournals/joss-eics should take over completely from here. I'm happy to back off, especially if my job is almost completely done anyway.

**Comment by danielskatz** [comment]
Given that the entire review history is public, I don't think that we need additional statements from reviewers. This hasn't been the practice of JOSS in the past.

**Comment by mstimberg** [comment]
> I appreciate you raising this, but I don't appreciate being told, even if indirectly, how to manage this.

I hear what you are saying, apologies for that – I'll try to be more careful in the future.

**Comment by oliviaguest** [comment]
@editorialbot recommend-accept

**Comment by danielskatz** [comment]
@editorialbot accept

**Comment by eferrea** [comment]
Great! @danielskatz @oliviaguest @mstimberg @puolival , many thanks for your contribution in this reviewing process.

**Comment by danielskatz** [comment]
The DOI doesn't yet resolve for me, so I'll keep this open until it does.

**Comment by danielskatz** [comment]
Congratulations to @eferrea (Enrico Ferrea) and co-authors!!

And thanks to @mstimberg and @puolival for reviewing, and @oliviaguest for editing!
We couldn't do this without you
